



<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      <meta http-equiv="x-ua-compatible" content="ie=edge">
      
        <meta name="description" content="csvtk - CSV/TSV toolkit">
      
      
      
        <meta name="author" content="shenwei356">
      
      
        <meta name="lang:clipboard.copy" content="Copy to clipboard">
      
        <meta name="lang:clipboard.copied" content="Copied to clipboard">
      
        <meta name="lang:search.language" content="en">
      
        <meta name="lang:search.pipeline.stopwords" content="True">
      
        <meta name="lang:search.pipeline.trimmer" content="True">
      
        <meta name="lang:search.result.none" content="No matching documents">
      
        <meta name="lang:search.result.one" content="1 matching document">
      
        <meta name="lang:search.result.other" content="# matching documents">
      
        <meta name="lang:search.tokenizer" content="[\s\-]+">
      
      <link rel="shortcut icon" href="files/favicon.ico">
      <meta name="generator" content="mkdocs-1.0.4, mkdocs-material-4.6.3">
    
    
      
        <title>csvtk - CSV/TSV Toolkit</title>
      
    
    
      <link rel="stylesheet" href="assets/stylesheets/application.adb8469c.css">
      
        <link rel="stylesheet" href="assets/stylesheets/application-palette.a8b3c06d.css">
      
      
        
        
        <meta name="theme-color" content="#03a9f4">
      
    
    
      <script src="assets/javascripts/modernizr.86422ebf.js"></script>
    
    
      
        <link href="https://fonts.gstatic.com" rel="preconnect" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,400,400i,700%7CRoboto+Mono&display=fallback">
        <style>body,input{font-family:"Roboto","Helvetica Neue",Helvetica,Arial,sans-serif}code,kbd,pre{font-family:"Roboto Mono","Courier New",Courier,monospace}</style>
      
    
    <link rel="stylesheet" href="assets/fonts/material-icons.css">
    
      <link rel="manifest" type="application/manifest+json" href="manifest.webmanifest" crossorigin="use-credentials">
    
    
    
      
        
<script>
  window.ga = window.ga || function() {
    (ga.q = ga.q || []).push(arguments)
  }
  ga.l = +new Date
  /* Setup integration and send page view */
  ga("create", "UA-28948626-9", "bioinf.shenwei.me/csvtk")
  ga("set", "anonymizeIp", true)
  ga("send", "pageview")
  /* Register handler to log search on blur */
  document.addEventListener("DOMContentLoaded", () => {
    if (document.forms.search) {
      var query = document.forms.search.query
      query.addEventListener("blur", function() {
        if (this.value) {
          var path = document.location.pathname;
          ga("send", "pageview", path + "?q=" + this.value)
        }
      })
    }
  })
</script>
<script async src="https://www.google-analytics.com/analytics.js"></script>
      
    
    
  </head>
  
    
    
    <body dir="ltr" data-md-color-primary="light-blue" data-md-color-accent="light-blue">
  
    <svg class="md-svg">
      <defs>
        
        
          <svg xmlns="http://www.w3.org/2000/svg" width="416" height="448" viewBox="0 0 416 448" id="__github"><path fill="currentColor" d="M160 304q0 10-3.125 20.5t-10.75 19T128 352t-18.125-8.5-10.75-19T96 304t3.125-20.5 10.75-19T128 256t18.125 8.5 10.75 19T160 304zm160 0q0 10-3.125 20.5t-10.75 19T288 352t-18.125-8.5-10.75-19T256 304t3.125-20.5 10.75-19T288 256t18.125 8.5 10.75 19T320 304zm40 0q0-30-17.25-51T296 232q-10.25 0-48.75 5.25Q229.5 240 208 240t-39.25-2.75Q130.75 232 120 232q-29.5 0-46.75 21T56 304q0 22 8 38.375t20.25 25.75 30.5 15 35 7.375 37.25 1.75h42q20.5 0 37.25-1.75t35-7.375 30.5-15 20.25-25.75T360 304zm56-44q0 51.75-15.25 82.75-9.5 19.25-26.375 33.25t-35.25 21.5-42.5 11.875-42.875 5.5T212 416q-19.5 0-35.5-.75t-36.875-3.125-38.125-7.5-34.25-12.875T37 371.5t-21.5-28.75Q0 312 0 260q0-59.25 34-99-6.75-20.5-6.75-42.5 0-29 12.75-54.5 27 0 47.5 9.875t47.25 30.875Q171.5 96 212 96q37 0 70 8 26.25-20.5 46.75-30.25T376 64q12.75 25.5 12.75 54.5 0 21.75-6.75 42 34 40 34 99.5z"/></svg>
        
      </defs>
    </svg>
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" data-md-component="overlay" for="__drawer"></label>
    
      <a href="#csvtk-a-cross-platform-efficient-and-practical-csvtsv-toolkit" tabindex="0" class="md-skip">
        Skip to content
      </a>
    
    
      <header class="md-header" data-md-component="header">
  <nav class="md-header-nav md-grid">
    <div class="md-flex">
      <div class="md-flex__cell md-flex__cell--shrink">
        <a href="." title="csvtk - CSV/TSV Toolkit" aria-label="csvtk - CSV/TSV Toolkit" class="md-header-nav__button md-logo">
          
            <i class="md-icon">school</i>
          
        </a>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        <label class="md-icon md-icon--menu md-header-nav__button" for="__drawer"></label>
      </div>
      <div class="md-flex__cell md-flex__cell--stretch">
        <div class="md-flex__ellipsis md-header-nav__title" data-md-component="title">
          
            <span class="md-header-nav__topic">
              csvtk - CSV/TSV Toolkit
            </span>
            <span class="md-header-nav__topic">
              
                Home
              
            </span>
          
        </div>
      </div>
      <div class="md-flex__cell md-flex__cell--shrink">
        
          <label class="md-icon md-icon--search md-header-nav__button" for="__search"></label>
          
<div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" aria-label="search" name="query" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="query" data-md-state="active">
      <label class="md-icon md-search__icon" for="__search"></label>
      <button type="reset" class="md-icon md-search__icon" data-md-component="reset" tabindex="-1">
        &#xE5CD;
      </button>
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" data-md-scrollfix>
        <div class="md-search-result" data-md-component="result">
          <div class="md-search-result__meta">
            Type to start searching
          </div>
          <ol class="md-search-result__list"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
        
      </div>
      
        <div class="md-flex__cell md-flex__cell--shrink">
          <div class="md-header-nav__source">
            


  

<a href="https://github.com/shenwei356/csvtk/" title="Go to repository" class="md-source" data-md-source="github">
  
    <div class="md-source__icon">
      <svg viewBox="0 0 24 24" width="24" height="24">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    GitHub
  </div>
</a>
          </div>
        </div>
      
    </div>
  </nav>
</header>
    
    <div class="md-container">
      
        
      
      
      <main class="md-main" role="main">
        <div class="md-main__inner md-grid" data-md-component="container">
          
            
              <div class="md-sidebar md-sidebar--primary" data-md-component="navigation">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    <nav class="md-nav md-nav--primary" data-md-level="0">
  <label class="md-nav__title md-nav__title--site" for="__drawer">
    <a href="." title="csvtk - CSV/TSV Toolkit" class="md-nav__button md-logo">
      
        <i class="md-icon">school</i>
      
    </a>
    csvtk - CSV/TSV Toolkit
  </label>
  
    <div class="md-nav__source">
      


  

<a href="https://github.com/shenwei356/csvtk/" title="Go to repository" class="md-source" data-md-source="github">
  
    <div class="md-source__icon">
      <svg viewBox="0 0 24 24" width="24" height="24">
        <use xlink:href="#__github" width="24" height="24"></use>
      </svg>
    </div>
  
  <div class="md-source__repository">
    GitHub
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
      

  


  <li class="md-nav__item md-nav__item--active">
    
    <input class="md-toggle md-nav__toggle" data-md-toggle="toc" type="checkbox" id="__toc">
    
      
    
    
      <label class="md-nav__link md-nav__link--active" for="__toc">
        Home
      </label>
    
    <a href="." title="Home" class="md-nav__link md-nav__link--active">
      Home
    </a>
    
      
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    Introduction
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    Features
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subcommands" class="md-nav__link">
    Subcommands
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    Installation
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#method-1-download-binaries-latest-stabledev-version" class="md-nav__link">
    Method 1: Download binaries (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-2-install-via-conda-latest-stable-version" class="md-nav__link">
    Method 2: Install via conda (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-3-for-go-developer-latest-stabledev-version" class="md-nav__link">
    Method 3: For Go developer (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-4-for-archlinux-aur-users-may-be-not-the-latest" class="md-nav__link">
    Method 4: For ArchLinux AUR users (may be not the latest)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bash-completion" class="md-nav__link">
    Bash-completion
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#compared-to-csvkit" class="md-nav__link">
    Compared to csvkit
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    Examples
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#acknowledgements" class="md-nav__link">
    Acknowledgements
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contact" class="md-nav__link">
    Contact
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#license" class="md-nav__link">
    License
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#starchart" class="md-nav__link">
    Starchart
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
    
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="download/" title="Download" class="md-nav__link">
      Download
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="usage/" title="Usage" class="md-nav__link">
      Usage
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="tutorial/" title="Tutorial" class="md-nav__link">
      Tutorial
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="chinese/" title="中文介绍" class="md-nav__link">
      中文介绍
    </a>
  </li>

    
      
      
      


  <li class="md-nav__item">
    <a href="https://github.com/shenwei356" title="More tools" class="md-nav__link">
      More tools
    </a>
  </li>

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              <div class="md-sidebar md-sidebar--secondary" data-md-component="toc">
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    
<nav class="md-nav md-nav--secondary">
  
  
    
  
  
    <label class="md-nav__title" for="__toc">Table of contents</label>
    <ul class="md-nav__list" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#introduction" class="md-nav__link">
    Introduction
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#table-of-contents" class="md-nav__link">
    Table of Contents
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#features" class="md-nav__link">
    Features
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#subcommands" class="md-nav__link">
    Subcommands
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#installation" class="md-nav__link">
    Installation
  </a>
  
    <nav class="md-nav">
      <ul class="md-nav__list">
        
          <li class="md-nav__item">
  <a href="#method-1-download-binaries-latest-stabledev-version" class="md-nav__link">
    Method 1: Download binaries (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-2-install-via-conda-latest-stable-version" class="md-nav__link">
    Method 2: Install via conda (latest stable version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-3-for-go-developer-latest-stabledev-version" class="md-nav__link">
    Method 3: For Go developer (latest stable/dev version)
  </a>
  
</li>
        
          <li class="md-nav__item">
  <a href="#method-4-for-archlinux-aur-users-may-be-not-the-latest" class="md-nav__link">
    Method 4: For ArchLinux AUR users (may be not the latest)
  </a>
  
</li>
        
      </ul>
    </nav>
  
</li>
      
        <li class="md-nav__item">
  <a href="#bash-completion" class="md-nav__link">
    Bash-completion
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#compared-to-csvkit" class="md-nav__link">
    Compared to csvkit
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#examples" class="md-nav__link">
    Examples
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#acknowledgements" class="md-nav__link">
    Acknowledgements
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#contact" class="md-nav__link">
    Contact
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#license" class="md-nav__link">
    License
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#starchart" class="md-nav__link">
    Starchart
  </a>
  
</li>
      
      
      
      
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          <div class="md-content">
            <article class="md-content__inner md-typeset">
              
                
                  <a href="https://github.com/shenwei356/csvtk/edit/master/docs/index.md" title="Edit this page" class="md-icon md-content__icon">&#xE3C9;</a>
                
                
                <h1 id="csvtk-a-cross-platform-efficient-and-practical-csvtsv-toolkit">csvtk - A cross-platform, efficient and practical CSV/TSV toolkit</h1>
<ul>
<li><strong>Documents:</strong> <a href="http://bioinf.shenwei.me/csvtk/">http://bioinf.shenwei.me/csvtk</a>
( <a href="http://bioinf.shenwei.me/csvtk/usage/"><strong>Usage</strong></a>  and <a href="http://bioinf.shenwei.me/csvtk/tutorial/"><strong>Tutorial</strong></a>). <a href="http://bioinf.shenwei.me/csvtk/chinese">中文介绍</a></li>
<li><strong>Source code:</strong>  <a href="https://github.com/shenwei356/csvtk">https://github.com/shenwei356/csvtk</a> <a href="https://github.com/shenwei356/csvtk"><img alt="GitHub stars" src="https://img.shields.io/github/stars/shenwei356/csvtk.svg?style=social&amp;label=Star&amp;?maxAge=2592000" /></a>
<a href="https://github.com/shenwei356/csvtk/blob/master/LICENSE"><img alt="license" src="https://img.shields.io/github/license/shenwei356/csvtk.svg?maxAge=2592000" /></a>
<a href="https://travis-ci.org/shenwei356/csvtk"><img alt="Build Status" src="https://travis-ci.org/shenwei356/csvtk.svg?branch=master" /></a></li>
<li><strong>Latest version:</strong> <a href="https://github.com/shenwei356/csvtk/releases"><img alt="Latest Stable Version" src="https://img.shields.io/github/release/shenwei356/csvtk.svg?style=flat" /></a>
<a href="http://bioinf.shenwei.me/csvtk/download/"><img alt="Github Releases" src="https://img.shields.io/github/downloads/shenwei356/csvtk/latest/total.svg?maxAge=3600" /></a>
<a href="http://bioinf.shenwei.me/csvtk/download/"><img alt="Cross-platform" src="https://img.shields.io/badge/platform-any-ec2eb4.svg?style=flat" /></a>
<a href="https://anaconda.org/bioconda/csvtk"><img alt="Anaconda Cloud" src="https://anaconda.org/bioconda/csvtk/badges/version.svg" /></a></li>
</ul>
<h2 id="introduction">Introduction</h2>
<p>Similar to FASTA/Q format in field of Bioinformatics,
CSV/TSV formats are basic and ubiquitous file formats in both Bioinformatics and data science.</p>
<p>People usually use spreadsheet software like MS Excel to process table data.
However this is all by clicking and typing, which is <strong>not
automated and is time-consuming to repeat</strong>, especially when you want to
apply similar operations with different datasets or purposes.</p>
<p><strong><em>You can also accomplish some CSV/TSV manipulations using shell commands,
but more code is needed to handle the header line. Shell commands do not
support selecting columns with column names either.</em></strong></p>
<p><code>csvtk</code> is <strong>convenient for rapid data investigation
and also easy to integrate into analysis pipelines</strong>.
It could save you lots of time in (not) writing Python/R scripts.</p>
<h2 id="table-of-contents">Table of Contents</h2>
<!-- START doctoc generated TOC please keep comment here to allow auto update -->

<!-- DON'T EDIT THIS SECTION, INSTEAD RE-RUN doctoc TO UPDATE -->

<ul>
<li><a href="#features">Features</a></li>
<li><a href="#subcommands">Subcommands</a></li>
<li><a href="#installation">Installation</a></li>
<li><a href="#bash-completion">Bash-completion</a></li>
<li><a href="#compared-to-csvkit">Compared to <code>csvkit</code></a></li>
<li><a href="#examples">Examples</a></li>
<li><a href="#acknowledgements">Acknowledgements</a></li>
<li><a href="#contact">Contact</a></li>
<li><a href="#license">License</a></li>
<li><a href="#starchart">Starchart</a></li>
</ul>
<!-- END doctoc generated TOC please keep comment here to allow auto update -->

<h2 id="features">Features</h2>
<ul>
<li><strong>Cross-platform</strong> (Linux/Windows/Mac OS X/OpenBSD/FreeBSD)</li>
<li><strong>Light weight and out-of-the-box, no dependencies, no compilation, no configuration</strong></li>
<li><strong>Fast</strong>,  <strong>multiple-CPUs supported</strong> (some commands)</li>
<li><strong>Practical functions provided by N subcommands</strong></li>
<li><strong>Support STDIN and gziped input/output file, easy being used in pipe</strong></li>
<li>Most of the subcommands support <strong><em>unselecting fields</em></strong> and <strong><em>fuzzy fields</em></strong>,
  e.g. <code>-f "-id,-name"</code> for all fields except "id" and "name",
  <code>-F -f "a.*"</code> for all fields with prefix "a.".</li>
<li><strong>Support some common plots</strong> (see <a href="http://bioinf.shenwei.me/csvtk/usage/#plot">usage</a>)</li>
<li><del>Seamlessly support for data with meta line (e.g., <code>sep=,</code>) of separator declaration used by MS Excel</del></li>
</ul>
<h2 id="subcommands">Subcommands</h2>
<p>45 subcommands in total.</p>
<p><strong>Information</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#headers"><code>headers</code></a>: prints headers</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#dim/nrow/ncol"><code>dim</code></a>: dimensions of CSV file</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#dim/nrow/ncol"><code>nrow</code></a>: print number of records</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#dim/nrow/ncol"><code>ncol</code></a>: print number of columns</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#summary"><code>summary</code></a>: summary statistics of selected digital fields (groupby group fields)</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#watch"><code>watch</code></a>: online monitoring and histogram of selected field</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#corr"><code>corr</code></a>: calculate Pearson correlation between numeric columns</li>
</ul>
<p><strong>Format conversion</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#pretty"><code>pretty</code></a>: converts CSV to readable aligned table</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#csv2tab"><code>csv2tab</code></a>: converts CSV to tabular format</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#tab2csv"><code>tab2csv</code></a>: converts tabular format to CSV</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#space2tab"><code>space2tab</code></a>: converts space delimited format to CSV</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#transpose"><code>transpose</code></a>: transposes CSV data</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#csv2md"><code>csv2md</code></a>: converts CSV to markdown format</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#csv2json"><code>csv2json</code></a>: converts CSV to JSON format</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#xlsx2csv"><code>xlsx2csv</code></a>: converts XLSX to CSV format</li>
</ul>
<p><strong>Set operations</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#head"><code>head</code></a>: prints first N records</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#concat"><code>concat</code></a>: concatenates CSV/TSV files by rows</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#sample"><code>sample</code></a>: sampling by proportion</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#cut"><code>cut</code></a>: selects parts of fields</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#grep"><code>grep</code></a>: greps data by selected fields with patterns/regular expressions</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#uniq"><code>uniq</code></a>: unique data without sorting</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#freq"><code>freq</code></a>: frequencies of selected fields</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#inter"><code>inter</code></a>: intersection of multiple files</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#filter"><code>filter</code></a>: filters rows by values of selected fields with arithmetic expression</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#filter2"><code>filter2</code></a>: filters rows by awk-like arithmetic/string expressions</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#join"><code>join</code></a>: join files by selected fields (inner, left and outer join)</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#split"><code>split</code></a> splits CSV/TSV into multiple files according to column values</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#splitxlsx"><code>splitxlsx</code></a>: splits XLSX sheet into multiple sheets according to column values</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#collapse"><code>collapse</code></a>: collapses one field with selected fields as keys</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#comb"><code>comb</code></a>: compute combinations of items at every row</li>
</ul>
<p><strong>Edit</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#add-header"><code>add-header</code></a>: add column names</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#del-header"><code>del-header</code></a>: delete column names</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#rename"><code>rename</code></a>: renames column names with new names</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#rename2"><code>rename2</code></a>: renames column names by regular expression</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#replace"><code>replace</code></a>: replaces data of selected fields by regular expression</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#round"><code>round</code></a>: round float to n decimal places</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#mutate"><code>mutate</code></a>: creates new columns from selected fields by regular expression</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#mutate2"><code>mutate2</code></a>: creates new column from selected fields by awk-like arithmetic/string expressions</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#sep"><code>sep</code></a>: separate column into multiple columns</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#gather"><code>gather</code></a>: gathers columns into key-value pairs</li>
</ul>
<p><strong>Ordering</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#sort"><code>sort</code></a>: sorts by selected fields</li>
</ul>
<p><strong>Ploting</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#plot"><code>plot</code></a> see <a href="http://bioinf.shenwei.me/csvtk/usage/#plot">usage</a><ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#hist"><code>plot hist</code></a> histogram</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#box"><code>plot box</code></a> boxplot</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#line"><code>plot line</code></a> line plot and scatter plot</li>
</ul>
</li>
</ul>
<p><strong>Misc</strong></p>
<ul>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#cat"><code>cat</code></a> stream file and report progress</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#version"><code>version</code></a>   print version information and check for update</li>
<li><a href="https://bioinf.shenwei.me/csvtk/usage/#genautocomplete"><code>genautocomplete</code></a> generate shell autocompletion script</li>
</ul>
<h2 id="installation">Installation</h2>
<p><a href="https://github.com/shenwei356/csvtk/releases">Download Page</a></p>
<p><code>csvtk</code> is implemented in <a href="https://golang.org/">Go</a> programming language,
 executable binary files <strong>for most popular operating systems</strong> are freely available
  in <a href="https://github.com/shenwei356/csvtk/releases">release</a> page.</p>
<h4 id="method-1-download-binaries-latest-stabledev-version">Method 1: Download binaries (latest stable/dev version)</h4>
<p>Just <a href="https://github.com/shenwei356/csvtk/releases">download</a> compressed
executable file of your operating system,
and decompress it with <code>tar -zxvf *.tar.gz</code> command or other tools.
And then:</p>
<ol>
<li>
<p><strong>For Linux-like systems</strong></p>
<ol>
<li>
<p>If you have root privilege simply copy it to <code>/usr/local/bin</code>:</p>
<pre><code>sudo cp csvtk /usr/local/bin/
</code></pre>
</li>
<li>
<p>Or copy to anywhere in the environment variable <code>PATH</code>:</p>
<pre><code>mkdir -p $HOME/bin/; cp csvtk $HOME/bin/
</code></pre>
</li>
</ol>
</li>
<li>
<p><strong>For windows</strong>, just copy <code>csvtk.exe</code> to <code>C:\WINDOWS\system32</code>.</p>
</li>
</ol>
<h4 id="method-2-install-via-conda-latest-stable-version">Method 2: Install via conda (latest stable version)  <a href="https://anaconda.org/bioconda/csvtk"><img alt="Anaconda Cloud" src="https://anaconda.org/bioconda/csvtk/badges/version.svg" /></a> <a href="https://anaconda.org/bioconda/csvtk"><img alt="downloads" src="https://anaconda.org/bioconda/csvtk/badges/downloads.svg" /></a></h4>
<pre><code>conda install -c bioconda csvtk
</code></pre>
<h4 id="method-3-for-go-developer-latest-stabledev-version">Method 3: For Go developer (latest stable/dev version)</h4>
<pre><code>go get -u github.com/shenwei356/csvtk/csvtk
</code></pre>
<h4 id="method-4-for-archlinux-aur-users-may-be-not-the-latest">Method 4: For ArchLinux AUR users (may be not the latest)</h4>
<pre><code>yaourt -S csvtk
</code></pre>
<h2 id="bash-completion">Bash-completion</h2>
<p>Note: The current version supports Bash only.
This should work for *nix systems with Bash installed.</p>
<p>Howto:</p>
<ol>
<li>
<p>run: <code>csvtk genautocomplete</code></p>
</li>
<li>
<p>create and edit <code>~/.bash_completion</code> file if you don't have it.</p>
<pre><code>nano ~/.bash_completion
</code></pre>
<p>add the following:</p>
<pre><code>for bcfile in ~/.bash_completion.d/* ; do
  . $bcfile
done
</code></pre>
</li>
</ol>
<h2 id="compared-to-csvkit">Compared to <code>csvkit</code></h2>
<p><a href="http://csvkit.readthedocs.org/">csvkit</a>, attention: this table wasn't updated for 2 years.</p>
<table>
<thead>
<tr>
<th align="left">Features</th>
<th align="center">csvtk</th>
<th align="center">csvkit</th>
<th align="left">Note</th>
</tr>
</thead>
<tbody>
<tr>
<td align="left">Read    Gzip</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="left">read gzip files</td>
</tr>
<tr>
<td align="left">Fields ranges</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="left">e.g. <code>-f 1-4,6</code></td>
</tr>
<tr>
<td align="left"><strong>Unselect fileds</strong></td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">e.g. <code>-1</code> for excluding first column</td>
</tr>
<tr>
<td align="left"><strong>Fuzzy fields</strong></td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">e.g. <code>ab*</code> for columns with name prefix "ab"</td>
</tr>
<tr>
<td align="left">Reorder fields</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="left">it means <code>-f 1,2</code> is different from <code>-f 2,1</code></td>
</tr>
<tr>
<td align="left"><strong>Rename columns</strong></td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">rename with new name(s) or from existed names</td>
</tr>
<tr>
<td align="left">Sort by multiple keys</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="left">bash sort like operations</td>
</tr>
<tr>
<td align="left"><strong>Sort by number</strong></td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">e.g. <code>-k 1:n</code></td>
</tr>
<tr>
<td align="left"><strong>Multiple sort</strong></td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">e.g. <code>-k 2:r -k 1:nr</code></td>
</tr>
<tr>
<td align="left">Pretty output</td>
<td align="center">Yes</td>
<td align="center">Yes</td>
<td align="left">convert CSV to readable aligned table</td>
</tr>
<tr>
<td align="left"><strong>Unique data</strong></td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">unique data of selected fields</td>
</tr>
<tr>
<td align="left"><strong>frequency</strong></td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">frequencies of selected fields</td>
</tr>
<tr>
<td align="left"><strong>Sampling</strong></td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">sampling by proportion</td>
</tr>
<tr>
<td align="left"><strong>Mutate fields</strong></td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">create new columns from selected fields</td>
</tr>
<tr>
<td align="left"><strong>Repalce</strong></td>
<td align="center">Yes</td>
<td align="center">--</td>
<td align="left">replace data of selected fields</td>
</tr>
</tbody>
</table>
<p>Similar tools:</p>
<ul>
<li><a href="http://csvkit.readthedocs.org/">csvkit</a> - A suite of utilities for converting to and working with CSV, the king of tabular file formats. http://csvkit.rtfd.org/</li>
<li><a href="https://github.com/BurntSushi/xsv">xsv</a> - A fast CSV toolkit written in Rust.</li>
<li><a href="https://github.com/johnkerl/miller">miller</a> - Miller is like sed, awk, cut, join, and sort for
name-indexed data such as CSV and tabular JSON http://johnkerl.org/miller</li>
<li><a href="https://github.com/eBay/tsv-utils">tsv-utils</a> - Command line utilities for tab-separated value files written in the D programming language.</li>
</ul>
<h2 id="examples">Examples</h2>
<p>More <a href="http://shenwei356.github.io/csvtk/usage/">examples</a> and <a href="http://shenwei356.github.io/csvtk/tutorial/">tutorial</a>.</p>
<p><strong>Attention</strong></p>
<ol>
<li>The CSV parser requires all the lines have same number of fields/columns.
    Even lines with spaces will cause error.
    Use '-I/--ignore-illegal-row' to skip these lines if neccessary.</li>
<li>By default, csvtk thinks your files have header row, if not, switch flag <code>-H</code> on.</li>
<li>Column names better be unique.</li>
<li>By default, lines starting with <code>#</code> will be ignored, if the header row
    starts with <code>#</code>, please assign flag <code>-C</code> another rare symbol, e.g. <code>'$'</code>.</li>
<li>By default, csvtk handles CSV files, use flag <code>-t</code> for tab-delimited files.</li>
<li>If <code>"</code> exists in tab-delimited files, use flag <code>-l</code>.</li>
<li>Do not mix use digital fields and column names.</li>
</ol>
<p>Examples</p>
<ol>
<li>
<p>Pretty result</p>
<pre><code>$ csvtk pretty names.csv
id   first_name   last_name   username
11   Rob          Pike        rob
2    Ken          Thompson    ken
4    Robert       Griesemer   gri
1    Robert       Thompson    abc
NA   Robert       Abel        123
</code></pre>
</li>
<li>
<p>Summary of selected digital fields, supporting "group-by"</p>
<pre><code>$ cat testdata/digitals2.csv \
    | csvtk summary --ignore-non-digits --fields f4:sum,f5:sum --groups f1,f2 \
    | csvtk pretty
f1    f2     f4:sum   f5:sum
bar   xyz    7.00     106.00
bar   xyz2   4.00     4.00
foo   bar    6.00     3.00
foo   bar2   4.50     5.00
</code></pre>
</li>
<li>
<p>Select fields/columns (<code>cut</code>)</p>
<ul>
<li>By index: <code>csvtk cut -f 1,2</code></li>
<li>By names: <code>csvtk cut -f first_name,username</code></li>
<li><strong>Unselect</strong>: <code>csvtk cut -f -1,-2</code> or <code>csvtk cut -f -first_name</code></li>
<li><strong>Fuzzy fields</strong>: <code>csvtk cut -F -f "*_name,username"</code></li>
<li>Field ranges: <code>csvtk cut -f 2-4</code> for column 2,3,4 or <code>csvtk cut -f -3--1</code> for discarding column 1,2,3</li>
<li>All fields: <code>csvtk cut -F -f "*"</code></li>
</ul>
</li>
<li>
<p>Search by selected fields (<code>grep</code>) (matched parts will be highlighted as red)</p>
<ul>
<li>By exactly matching: <code>csvtk grep -f first_name -p Robert -p Rob</code></li>
<li>By regular expression: <code>csvtk grep -f first_name -r -p Rob</code></li>
<li>By pattern list: <code>csvtk grep -f first_name -P name_list.txt</code></li>
<li>Remore rows containing missing data (NA): <code>csvtk grep -F -f "*" -r -p "^$" -v</code></li>
</ul>
</li>
<li>
<p><strong>Rename column names</strong> (<code>rename</code> and <code>rename2</code>)</p>
<ul>
<li>Setting new names: <code>csvtk rename -f A,B -n a,b</code> or <code>csvtk rename -f 1-3 -n a,b,c</code></li>
<li>Replacing with original names by regular express: <code>cat ../testdata/c.csv | ./csvtk rename2 -F -f "*" -p "(.*)" -r 'prefix_$1'</code> for adding prefix to all column names.</li>
</ul>
</li>
<li>
<p><strong>Edit data with regular expression</strong> (<code>replace</code>)</p>
<ul>
<li>Remove Chinese charactors:  <code>csvtk replace -F -f "*_name" -p "\p{Han}+" -r ""</code></li>
</ul>
</li>
<li>
<p><strong>Create new column from selected fields by regular expression</strong> (<code>mutate</code>)</p>
<ul>
<li>In default, copy a column: <code>csvtk mutate -f id</code></li>
<li>Extract prefix of data as group name (get "A" from "A.1" as group name):
  <code>csvtk mutate -f sample -n group -p "^(.+?)\."</code></li>
</ul>
</li>
<li>
<p>Sort by multiple keys (<code>sort</code>)</p>
<ul>
<li>By single column : <code>csvtk sort -k 1</code> or <code>csvtk sort -k last_name</code></li>
<li>By multiple columns: <code>csvtk sort -k 1,2</code> or <code>csvtk sort -k 1 -k 2</code> or <code>csvtk sort -k last_name,age</code></li>
<li>Sort by number: <code>csvtk sort -k 1:n</code> or  <code>csvtk sort -k 1:nr</code> for reverse number</li>
<li>Complex sort: <code>csvtk sort -k region -k age:n -k id:nr</code></li>
<li>In natural order: <code>csvtk sort -k chr:N</code></li>
</ul>
</li>
<li>
<p><strong>Join multiple files by keys</strong> (<code>join</code>)</p>
<ul>
<li>All files have same key column: <code>csvtk join -f id file1.csv file2.csv</code></li>
<li>Files have different key columns: <code>csvtk join -f "username;username;name" names.csv phone.csv adress.csv -k</code></li>
</ul>
</li>
<li>
<p>Filter by numbers (<code>filter</code>)</p>
<ul>
<li>Single field: <code>csvtk filter -f "id&gt;0"</code></li>
<li><strong>Multiple fields</strong>: <code>csvtk filter -f "1-3&gt;0"</code></li>
<li>Using <code>--any</code> to print record if any of the field satisfy the condition: <code>csvtk filter -f "1-3&gt;0" --any</code></li>
<li><strong>fuzzy fields</strong>: <code>csvtk filter -F -f "A*!=0"</code></li>
</ul>
</li>
<li>
<p><strong>Filter rows by awk-like arithmetic/string expressions</strong> (<code>filter2</code>)</p>
<ul>
<li>Using field index: <code>csvtk filter2 -f '$3&gt;0'</code></li>
<li>Using column names: <code>csvtk filter2 -f '$id &gt; 0'</code></li>
<li>Both arithmetic and string expressions: <code>csvtk filter2 -f '$id &gt; 3 || $username=="ken"'</code></li>
<li>More complicated: <code>csvtk filter2 -H -t -f '$1 &gt; 2 &amp;&amp; $2 % 2 == 0'</code></li>
</ul>
</li>
<li>
<p>Ploting</p>
<ul>
<li>plot histogram with data of the second column:<pre><code>csvtk -t plot hist testdata/grouped_data.tsv.gz -f 2 | display
</code></pre>
</li>
</ul>
<p><img alt="histogram.png" src="testdata/figures/histogram.png" /></p>
<ul>
<li>plot boxplot with data of the "GC Content" (third) column,
group information is the "Group" column.<pre><code>csvtk -t plot box testdata/grouped_data.tsv.gz -g "Group" \
    -f "GC Content" --width 3 | display
</code></pre>
</li>
</ul>
<p><img alt="boxplot.png" src="testdata/figures/boxplot.png" /></p>
<ul>
<li>plot horiz boxplot with data of the "Length" (second) column,
group information is the "Group" column.<pre><code>csvtk -t plot box testdata/grouped_data.tsv.gz -g "Group" -f "Length"  \
    --height 3 --width 5 --horiz --title "Horiz box plot" | display
</code></pre>
</li>
</ul>
<p><img alt="boxplot2.png" src="testdata/figures/boxplot2.png" /></p>
<ul>
<li>plot line plot with X-Y data<pre><code>csvtk -t plot line testdata/xy.tsv -x X -y Y -g Group | display
</code></pre>
</li>
</ul>
<p><img alt="lineplot.png" src="testdata/figures/lineplot.png" /></p>
<ul>
<li>plot scatter plot with X-Y data<pre><code>csvtk -t plot line testdata/xy.tsv -x X -y Y -g Group --scatter | display
</code></pre>
</li>
</ul>
<p><img alt="scatter.png" src="testdata/figures/scatter.png" /></p>
</li>
</ol>
<h2 id="acknowledgements">Acknowledgements</h2>
<p>We are grateful to <a href="https://github.com/dawnmy">Zhiluo Deng</a> and
<a href="https://github.com/penglbio">Li Peng</a> for suggesting features and reporting bugs.</p>
<p>Thanks <a href="https://github.com/avilella">Albert Vilella</a> for features suggestion,
which makes csvtk feature-rich。</p>
<h2 id="contact">Contact</h2>
<p><a href="https://github.com/shenwei356/csvtk/issues">Create an issue</a> to report bugs,
propose new functions or ask for help.</p>
<p>Or <a href="https://shenwei356.github.io/csvtk/usage/#disqus_thread">leave a comment</a>.</p>
<h2 id="license">License</h2>
<p><a href="https://github.com/shenwei356/csvtk/blob/master/LICENSE">MIT License</a></p>
<h2 id="starchart">Starchart</h2>
<p><img src="https://starchart.cc/shenwei356/csvtk.svg" alt="Stargazers over time" style="max-width: 100%"></p>
                
                  
                
                
              
              
                


              
            </article>
          </div>
        </div>
      </main>
      
        
<footer class="md-footer">
  
    <div class="md-footer-nav">
      <nav class="md-footer-nav__inner md-grid">
        
        
          <a href="download/" title="Download" class="md-flex md-footer-nav__link md-footer-nav__link--next" rel="next">
            <div class="md-flex__cell md-flex__cell--stretch md-footer-nav__title">
              <span class="md-flex__ellipsis">
                <span class="md-footer-nav__direction">
                  Next
                </span>
                Download
              </span>
            </div>
            <div class="md-flex__cell md-flex__cell--shrink">
              <i class="md-icon md-icon--arrow-forward md-footer-nav__button"></i>
            </div>
          </a>
        
      </nav>
    </div>
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-footer-copyright">
        
        powered by
        <a href="https://www.mkdocs.org" target="_blank" rel="noopener">MkDocs</a>
        and
        <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
          Material for MkDocs</a>
      </div>
      
    </div>
  </div>
</footer>
      
    </div>
    
      <script src="assets/javascripts/application.c33a9706.js"></script>
      
      <script>app.initialize({version:"1.0.4",url:{base:"."}})</script>
      
    
  </body>
</html>